### PR TITLE
Add resource-based job throttling to activejob limiter

### PR DESCRIPTION
**Pull request by andrewhamon**
This PR adds a new `throttle_job` macro which deduplicates jobs by resource and tries to have good latency and correctnes at the same time. It does this by using locks at both enqueue-time and perform-time.

An example of a solution that would have bad latency but good correctness: artifically delaying execution by `duration`, and disallow duplicates in the queue. This will cause a delay in all cases, which might not be desirable if a user is waiting.

An example of a solution that would have good latency but bad correctness: dropping all jobs if they are enqueued within `duration` of the start of an identical job. Sometimes it is possible that seperate changes do happen with relatively high frequency, and they might not be "seen" by the first job.

How this PR does it:

At enqueue time, we employ near-identical techniques as `limit_queue`. That is, a job holds a lock from the time it is enqueued until right before it is performed. This can help prevent duplicate jobs that occur very close together, but if the queue is empty, then a job will spend an extremely small amount of time in this period. So it is not a concurrency gaurantee by any means, but is a handy way to drop duplicate jobs without worring about "losing" anything (assuming the job is idempotent).

At perform time, a lock is acquired and never released. Instead it expires after the specified duration. As long as the job doesn't run for longer than the specified duration, this should provide a pretty good gaurantee that a job will never run more frequently than once every `duration`.

The behavior for subsequent duplicate jobs at perform time (which can occur if the job is relatively slow to run -- as soon as it leaves the queue, an equivelant job can enter they queue, and attempt to execute at perform time) is as follows:

- on the first dupe, another lock is acquired, known as the reschedule lock. The job is not performed, but instead rescheduled for a future time (`1.1 * duration`). The lock is still only held for `duration` though.
- every subsequent duplicate job is dropped. We can do this with confidence, since we knows that an identical job has been scheduled for the near future.

**tl;dr: this method will only drop a job if it is gauranteed that an equivelent job is enqueued but has not yet began execution. It artificially enqueues future jobs so that this is true more often. The worst case scenario is that n duplicate jobs will result in either 1 or 2 job invocations**.

To use, add something like this to your job class:

```ruby
throttle_job(
      duration: 2.minutes,
      extract_resource_id: (lambda { |job|
        job.arguments.first
      }),
```

There is an optional argument, metrics_hook, which will be fired and provided with information about what happened to the job (e.g. was it dropped at enqueue time? Was it enqueued, but dropped at perform time? etc.) I intend to use this to monitor a production roll out.

### CODE REVIEW

File: spec/active_job/limiter_spec.rb [hunk 1]
```diff
@@ -50,7 +50,7 @@ def perform; end
         .with(instance_of(LimitedJob), expiration_time).and_return(false)
       expect(ActiveJob::Limiter).to_not receive(:clear_lock_before_perform)
       job = LimitedJob.perform_later
-      expect(job.job_id).to be_nil
+      expect(job == false || job.job_id == nil).to be true
```
Comment (andrewhamon) [inline spec/active_job/limiter_spec.rb]: Tests were failing because of this. I guess perform_later stopped retuirning the job class?

File: lib/active_job/limiter/mixin.rb [hunk 2]
```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
...
+              metrics_hook.call('enqueue.dropped', job)
+            end
+          end
+        end
+
+        def active_job_limiter_add_throttle_job_around_perform(duration:, extract_resource_id:, metrics_hook:)
+          around_perform do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if ActiveJob::Limiter.acquire_lock_for_job_resource('perform', duration, job, resource_id)
```
Comment (dpkp) [inline lib/active_job/limiter/mixin.rb:92]: when is this lock released?

File: lib/active_job/limiter/mixin.rb [hunk 3]
```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
...
+              ActiveJob::Limiter.release_lock_for_job_resource('enqueue', job, resource_id)
+              block.call
+              metrics_hook.call('perform.performed', job)
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('reschedule', duration, job, resource_id)
```
Comment (dpkp) [inline lib/active_job/limiter/mixin.rb:95]: when is this lock released?

File: lib/active_job/limiter/mixin.rb [hunk 4]
```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
...
+
+        # Prefixing private methods with active_job_limiter to avoid conflicts with user code
+
+        def active_job_limiter_add_throttle_job_around_enqueue(duration:, extract_resource_id:, metrics_hook:)
+          around_enqueue do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if job.instance_variable_get(:@bypass_active_job_limiter_enqueue_locks)
+              block.call
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('enqueue', duration, job, resource_id)
```
Comment (dpkp) [inline lib/active_job/limiter/mixin.rb]: possible that we acquire lock but crash before enqueue operation...

File: lib/active_job/limiter/mixin.rb [hunk 5]
```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
...
+              job.job_id = nil
+              metrics_hook.call('perform.dropped', job)
+            end
+          end
+        end
+
+        def active_job_limiter_reschedule_job_for_later(existing_job, lock_duration)
+          new_job = existing_job.class.new(*existing_job.arguments)
+
+          # Rescheduled jobs need to pass through around_enqueue without being dropped, so we
```
Comment (dpkp) [inline lib/active_job/limiter/mixin.rb]: why is this?

Comment (andrewhamon) [reply lib/active_job/limiter/mixin.rb]: If we get to this point, that means a different job has already acquired the enqueue lock and the perform lock. When we re-enqueue this job, activejob will once again run it through the around_enqueue hook. So if we don’t set up an escape hatch then this enqueue would always be dropped.

A consequence of this is that even for end users, it is impossible to schedule a job for the future if there is an equivalent job currently enqueued. The original limit_queue also suffers from this flaw. I don’t see it has high priority since it is quite rare for us to schedule future sidekiq jobs at Flexport, and I imagine it would be even more rare that we would do something like that for the kinds of high frequency jobs that this library is meant to help with.

File: lib/active_job/limiter/mixin.rb (hunk 4 above)
Comment (andrewhamon) [reply lib/active_job/limiter/mixin.rb]: In that case Sidekiq should retry the job, right? Which should re-run the around_perform hook, and lead to a reschedule to occur (since the perform lock will have been acquired).

Of course, the server could also crash between getting the reschedule lock and scheduling the job for the future.

That would be a once-in-many-billion event, though. To me, that is not worth diving into what I’m sure would be activejob/sidekiq internals to try to atomically lock-and-schedule. limit_queue has been serving us well for years without this feature.

File: lib/active_job/limiter/mixin.rb (hunk 3 above)
Comment (andrewhamon) [reply lib/active_job/limiter/mixin.rb:95]: It is automatically released by redis after the specified duration (it adds an expire to the key).

File: lib/active_job/limiter/mixin.rb (hunk 2 above)
Comment (andrewhamon) [reply lib/active_job/limiter/mixin.rb:92]: See my other comment, it is never explicitly released, but automatically released by redis by setting an expiration on the key.

**Review by dpkp — APPROVED** [review]
